torch/jit/_recursive.py

```diff
@@ -291,7 +305,7 @@ def create_methods_from_stubs(concrete_type, stubs):
     defaults = [get_default_args(m.original_method) for m in stubs]
     concrete_type._create_methods(defs, rcbs, defaults)
 
-def create_script_module(nn_module, stubs_fn, share_types=True):
+def create_script_module(nn_module, stubs_fn, submodule_stubs_fn=None, share_types=True):
```

Comment:
Do we need to add this extra argument? Is there ever going to be a case where we don't use all the functions that are marked with `_jit_internal.FunctionModifiers.EXPORT` ?

Reply:
Before this PR, the rule is: submodule will be compiled using the default inference rule, which only contains the `forward` method + export methods, so export methods will always be compiled before/after this PR. This argument is added because we don't want tracing to compile the `forward` method, we want the tracing to compile the export methods, but still trace the forward calls

Reply:
Isn't this what `stubs_fn` is already for ?

Reply:
`stubs_fn` is only used for the current module compilation, when we are doing recursive compilation for the current module's submodule, we are using `infer_methods_to_compile` as the `stubs_fn`, which disallow us from compiling the submodule. I was trying to also use the `stubs_fn` as the rule for submodule compilation to replace `infer_methods_to_compile`, but it failed because we need to maintain the legacy `ScriptModule` api where submodule don't have `cls._methods` available. See https://github.com/pytorch/pytorch/blob/7816d53798efcd1dc68a83a2964ee6d4ef4181ca/torch/jit/_script.py#L203

Reply:
Copied from chat: 

```
we are creating ScriptModule using make_stubs function, and the current ScriptModule initialization use cls._methods as stubs_fn infer rule, but that ScriptModule's submodule is compiled using recursive compilation, if we just use stubs_fn as the submodule method infer rule, the submodule compilation also needs cls._methods, which is not available as cls._methods only available in the current ScriptModule (we construct cls._methods in ScriptMeta.__init__
```